src/node/history.h

```diff
@@ -490,13 +500,80 @@ namespace ccf
       NodeId id_,
       tls::KeyPair& kp_,
       Signatures& sig_,
-      Nodes& nodes_) :
+      Nodes& nodes_,
+      size_t sig_tx_interval_ = 0,
+      size_t sig_ms_interval_ = 0) :
       store(store_),
       id(id_),
       kp(kp_),
       signatures(sig_),
-      nodes(nodes_)
-    {}
+      nodes(nodes_),
+      sig_tx_interval(sig_tx_interval_),
+      sig_ms_interval(sig_ms_interval_)
+    {
+      start_signature_emit_timer();
+    }
+
+    void start_signature_emit_timer()
+    {
+      struct EmitSigMsg
+      {
+        EmitSigMsg(HashedTxHistory<T>* self_) : self(self_) {}
+        HashedTxHistory<T>* self;
+      };
+
+      auto emit_sig_msg = std::make_unique<threading::Tmsg<EmitSigMsg>>(
+        [](std::unique_ptr<threading::Tmsg<EmitSigMsg>> msg) {
+          auto self = msg->data.self;
+
+          std::unique_lock<SpinLock> mguard(
+            self->signature_lock, std::defer_lock);
+
+          const int64_t sig_ms_interval = self->sig_ms_interval;
+          int64_t time_since_last_sig = sig_ms_interval;
+
+          if (mguard.try_lock())
+          {
+            auto time = threading::ThreadMessaging::thread_messaging
+                          .get_current_time_offset();
+
+            auto consensus = self->store.get_consensus();
+            if (
+              (consensus != nullptr) && consensus->is_primary() &&
+              self->store.commit_gap() > 0 &&
+              time.count() > self->time_of_last_signature.count() &&
+              (time.count() - self->time_of_last_signature.count()) >
+                sig_ms_interval)
+            {
+              msg->data.self->emit_signature();
+            }
+
+            time_since_last_sig =
```

Comment:
This isn't time since last sig? (time - self->time_of_last_signature) is time since last sig or am I missing something?

Reply:
I guess that's time_to_next_sig? I don't understand why it's not sig_msg_interval every time though.

Reply:
This function is called every sig_msg_interval but we could have signed in between callbacks. We will therefore skip this signature (line 545) and so we do not want to schedule the next signature to be sig_msg_interval after now but rather after the last signature. 

Please confirm that my logic is correct because i could be really wrong here.

Reply:
That's fine logic-wise, but the name is confusing is. Time since last sig is `time - self->time_of_last_signature.count()`.

`sig_ms_interval - time_since_last_sig` should be called `time_to_next_sig`.

Reply:
time_to_next_sig